File: src-tauri/src/commands/security.rs
use tauri::command;

use crate::services::security::{
    ApprovalPolicy, ApprovalRecord, PermissionRules, RuleAction, RuleDecision, SandboxMode,
    SecurityConfig, SecurityService,
};

/// 读取当前安全配置
#[command]
pub fn get_security() -> SecurityConfig {
    SecurityService::load()
}

/// 设置沙箱模式（read-only / workspace-write / danger-full-access）
#[command]
pub fn set_sandbox_mode(mode: String) -> Result<SecurityConfig, String> {
    let parsed = match mode.as_str() {
        "read-only" => SandboxMode::ReadOnly,
        "workspace-write" => SandboxMode::WorkspaceWrite,
        "danger-full-access" => SandboxMode::DangerFullAccess,
        _ => return Err(format!("Unknown sandbox mode: {}", mode)),
    };
    SecurityService::set_sandbox_mode(parsed).map_err(|e| e.to_string())
}

/// 设置审批策略（untrusted / on-request / never）
#[command]
pub fn set_approval_policy(policy: String) -> Result<SecurityConfig, String> {
    let parsed = match policy.as_str() {
        "untrusted" => ApprovalPolicy::Untrusted,
        "on-request" => ApprovalPolicy::OnRequest,
        "never" => ApprovalPolicy::Never,
        _ => return Err(format!("Unknown approval policy: {}", policy)),
    };
    SecurityService::set_approval_policy(parsed).map_err(|e| e.to_string())
}

/// 设置 git 自动快照开关（turn/completed 后自动 commit 本地快照）
#[command]
pub fn set_auto_checkpoint(enabled: bool) -> Result<SecurityConfig, String> {
    SecurityService::set_auto_checkpoint(enabled).map_err(|e| e.to_string())
}

/// 记录一条审批历史
#[command]
pub fn record_approval(
    command: String,
    approved: bool,
    run_id: String,
) -> Result<SecurityConfig, String> {
    let record = ApprovalRecord {
        id: format!(
            "ap-{}",
            std::time::SystemTime::now()
                .duration_since(std::time::UNIX_EPOCH)
                .map(|d| d.as_millis())
                .unwrap_or(0)
        ),
        timestamp: std::time::SystemTime::now()
            .duration_since(std::time::UNIX_EPOCH)
            .map(|d| d.as_secs() as i64)
            .unwrap_or(0),
        command,
        approved,
        run_id,
    };
    SecurityService::add_history(record).map_err(|e| e.to_string())
}

/// 清空审批历史
#[command]
pub fn clear_approval_history() -> Result<SecurityConfig, String> {
    SecurityService::clear_history().map_err(|e| e.to_string())
}

/// 新增权限规则（deny/allow）
#[command]
pub fn add_permission_rule(
    pattern: String,
    action: String,
    note: Option<String>,
) -> Result<PermissionRules, String> {
    let parsed = match action.as_str() {
        "deny" => RuleAction::Deny,
        "allow" => RuleAction::Allow,
        _ => return Err(format!("Unknown rule action: {}", action)),
    };
    SecurityService::add_rule(pattern, parsed, note.unwrap_or_default()).map_err(|e| e.to_string())
}

/// 删除权限规则（按 index）
#[command]
pub fn remove_permission_rule(index: usize) -> Result<PermissionRules, String> {
    SecurityService::remove_rule(index).map_err(|e| e.to_string())
}

/// 读取全部权限规则
#[command]
pub fn list_permission_rules() -> PermissionRules {
    SecurityService::load_rules()
}

/// 清空权限规则
#[command]
pub fn clear_permission_rules() -> Result<PermissionRules, String> {
    SecurityService::clear_rules().map_err(|e| e.to_string())
}

/// 规则决策测试结果（7.3 GUI 实测入口，不执行命令，仅返回规则引擎判定）
#[derive(serde::Serialize)]
pub struct RuleTestResult {
    /// auto_deny / auto_accept / ask
    pub tag: String,
    pub reason: String,
    pub denied: bool,
}

/// 输入任意命令，返回规则引擎的决策结果（deny/allow/ask），用于 GUI 实测与规则校验。
/// 只判定、不执行；模型自觉拒绝对此无影响（这是纯规则层验证入口）。
#[command]
pub fn security_test_rule(command: String) -> RuleTestResult {
    let d = SecurityService::decide(&command);
    match d {
        RuleDecision::BuiltinDeny(r) => RuleTestResult { tag: "auto_deny".into(), reason: r, denied: true },
        RuleDecision::UserDeny(r) => RuleTestResult { tag: "auto_deny".into(), reason: r, denied: true },
        RuleDecision::UserAllow(r) => RuleTestResult { tag: "auto_accept".into(), reason: r, denied: false },
        RuleDecision::AutoAllow => RuleTestResult { tag: "auto_accept".into(), reason: "全自动模式（approval_policy=never）放行".into(), denied: false },
        RuleDecision::Ask => RuleTestResult { tag: "ask".into(), reason: "需要审批（未命中 deny/allow 规则）".into(), denied: false },
    }
}

File: src/types/security.ts
/** 沙箱模式（对应 codex sandbox_mode） */
export type SandboxMode = 'read-only' | 'workspace-write' | 'danger-full-access'

/** 审批策略（对应 codex approval_policy） */
export type ApprovalPolicy = 'untrusted' | 'on-request' | 'never'

/** 审批历史记录 */
export interface ApprovalRecord {
  id: string
  timestamp: number
  command: string
  approved: boolean
  run_id: string
}

/** 权限规则动作（deny 优先于 allow） */
export type RuleAction = 'deny' | 'allow'

/** 单条权限规则 */
export interface PermissionRule {
  pattern: string
  action: RuleAction
  note: string
}

/** 权限规则集（持久化 ~/.flydex/permissions.json） */
export interface PermissionRules {
  rules: PermissionRule[]
}

/** 规则决策测试结果（7.3 security_test_rule） */
export interface RuleTestResult {
  tag: 'auto_deny' | 'auto_accept' | 'ask'
  reason: string
  denied: boolean
}

/** 决策标签展示元信息 */
export const RULE_TEST_TAGS: {
  value: RuleTestResult['tag']
  label: string
  color: string
  desc: string
}[] = [
  {
    value: 'auto_deny',
    label: '自动拒绝',
    color: 'text-red-500 bg-red-500/10',
    desc: '规则引擎判定危险，工具执行前直接拒绝',
  },
  {
    value: 'auto_accept',
    label: '自动放行',
    color: 'text-emerald-500 bg-emerald-500/10',
    desc: '命中 allow 规则或全自动模式，直接执行',
  },
  {
    value: 'ask',
    label: '需审批',
    color: 'text-amber-500 bg-amber-500/10',
    desc: '未命中 deny/allow 规则，推审批框由你决定',
  },
]

export function ruleTestTagLabel(tag: RuleTestResult['tag']): string {
  return RULE_TEST_TAGS.find((t) => t.value === tag)?.label ?? tag
}

/** 安全配置 */
export interface SecurityConfig {
  sandbox_mode: SandboxMode
  approval_policy: ApprovalPolicy
  /** git 自动快照开关：每轮 turn 完成后自动 commit 本地快照（对齐 Claude Code） */
  auto_checkpoint: boolean
  history: ApprovalRecord[]
}

/** 沙箱模式展示元信息 */
export const SANDBOX_MODES: { value: SandboxMode; label: string; desc: string; icon: string }[] = [
  {
    value: 'read-only',
    label: '只读',
    desc: 'AI 只能读文件，无法写入或执行写操作。最安全，适合纯分析。',
    icon: '👁️',
  },
  {
    value: 'workspace-write',
    label: '工作区写入',
    desc: '允许在项目目录内写入和修改。⚠️ Windows 上对 .git 目录写入受限，git add/commit 可能失败，建议改用完全访问完成 git 操作。',
    icon: '📝',
  },
  {
    value: 'danger-full-access',
    label: '完全访问',
    desc: '无沙箱，AI 以你的真实权限运行，可操作任意文件。高风险。',
    icon: '⚡',
  },
]

/** 审批策略展示元信息 */
export const APPROVAL_POLICIES: {
  value: ApprovalPolicy
  label: string
  desc: string
  icon: string
}[] = [
  {
    value: 'untrusted',
    label: '询问不可信操作',
    desc: '只对 codex 判定为不可信的命令请求审批，常规操作自动执行。',
    icon: '🛡️',
  },
  {
    value: 'on-request',
    label: '按需请求',
    desc: '模型在需要提权或执行敏感操作时主动请求你确认（推荐）。',
    icon: '🤝',
  },
  {
    value: 'never',
    label: '自动放行',
    desc: '不询问任何操作，AI 全自动执行。仅适合可信环境。',
    icon: '🚀',
  },
]

export function sandboxLabel(mode: SandboxMode): string {
  return SANDBOX_MODES.find((m) => m.value === mode)?.label ?? mode
}

export function approvalLabel(policy: ApprovalPolicy): string {
  return APPROVAL_POLICIES.find((p) => p.value === policy)?.label ?? policy
}
